op-supervisor/supervisor/backend/cross/hazard_set.go

```diff
@@ -0,0 +1,178 @@
+		if len(execMsgs) > 0 {
+			if ok, err := depSet.CanExecuteAt(next.chainID, candidate.Timestamp); err != nil {
+				return fmt.Errorf("cannot check message execution of block %s (chain %s): %w", candidate, next.chainID, err)
+			} else if !ok {
+				return fmt.Errorf("cannot execute messages in block %s (chain %s): %w", candidate, next.chainID, types.ErrConflict)
+			}
+		}
+
+		// check all executing messages
+		for _, msg := range execMsgs {
+			logger.Debug("Processing executing message", "chainID", next.chainID, "block", candidate, "msg", msg)
+
+			initChainID, err := depSet.ChainIDFromIndex(msg.Chain)
+			if err != nil {
+				if errors.Is(err, types.ErrUnknownChain) {
+					err = fmt.Errorf("msg %s may not execute from unknown chain %s: %w", msg, msg.Chain, types.ErrConflict)
+				}
+				return err
+			}
+
+			// invariant: the chain must be able to initiate at the timestamp of the message we're referencing
+			if ok, err := depSet.CanInitiateAt(initChainID, msg.Timestamp); err != nil {
+				return fmt.Errorf("cannot check message initiation of msg %s (chain %s): %w", msg, chainID, err)
+			} else if !ok {
+				return fmt.Errorf("cannot allow initiating message %s (chain %s): %w", msg, chainID, types.ErrConflict)
+			}
+
+			// invariant: the message must not be in the future
+			if msg.Timestamp > candidate.Timestamp {
+				return fmt.Errorf("executing message %s in %s breaks timestamp invariant", msg, candidate)
+			}
+
+			includedIn, err := deps.Contains(initChainID,
+				types.ContainsQuery{
+					Timestamp: msg.Timestamp,
+					BlockNum:  msg.BlockNum,
+					LogIdx:    msg.LogIdx,
+					LogHash:   msg.Hash,
+				})
+			if err != nil {
+				return fmt.Errorf("executing msg %s failed check: %w", msg, err)
+			}
+
+			if msg.Timestamp < candidate.Timestamp {
+				// If timestamp is older: invariant ensures non-cyclic ordering relative to other messages.
+				// Ensure that the block that they are included in is cross-safe.
+				if err := deps.IsCrossValidBlock(initChainID, includedIn.ID()); err != nil {
+					return fmt.Errorf("msg %s included in non-cross-safe block %s: %w", msg, includedIn, err)
+				}
+			} else if msg.Timestamp == candidate.Timestamp {
+				// If timestamp is equal: we have to inspect ordering of individual
+				// log events to ensure non-cyclic cross-chain message ordering.
+				// And since we may have back-and-forth messaging, we cannot wait till the initiating side is cross-safe.
+				// Thus check that it was included in a local-safe block,
+				// and then proceed with transitive block checks,
+				// to ensure the local block we depend on is becoming cross-safe also.
+				logger.Debug("Checking message with current timestamp", "msg", msg, "candidate", candidate)
+
+				if existing, ok := h.entries[msg.Chain]; ok {
+					if existing.ID() != includedIn.ID() {
+						return fmt.Errorf("found dependency on %s (chain %d), but already depend on %s", includedIn, initChainID, chainID)
```

Comment:
Same with these stanzas, depending on the timestamps, there is different processing done, and each can return errors. Could these be factored into helper methods?